mu/virtual_environment.py

```diff
@@ -0,0 +1,553 @@
+        #
+        if (
+            not finished
+            and self.process.exitStatus() == self.process.CrashExit
+        ):
+            raise RuntimeError("Some error occurred")
+
+    def data(self):
+        return self.process.readAll().data().decode("utf-8")
+
+    def _started(self):
+        self.started.emit()
+
+    def _readyRead(self):
+        self.output.emit(self.data().strip())
+
+    def _finished(self):
+        self.finished.emit()
+
+
+class Pip(object):
+    """Proxy for various pip commands
+
+    While this is a fairly useful abstraction in its own right, it's at
+    least initially to assist in testing, so we can mock out various
+    commands
+    """
+
+    def __init__(self, pip_executable):
+        self.executable = pip_executable
+        self.process = Process()
+
+    def run(
+        self, command, *args, wait_for_s=30.0, slots=Process.Slots(), **kwargs
+    ):
+        """Run a command with args, treating kwargs as Posix switches
+
+        eg run("python", version=True)
+        run("python", "-c", "import sys; print(sys.executable)")
+        """
+        #
+        # Any keyword args are treated as command-line switches
+        # As a special case, a boolean value indicates that the flag
+        # is a yes/no switch
+        #
+        params = [command, "--disable-pip-version-check"]
+        for k, v in kwargs.items():
+            switch = k.replace("_", "-")
+            if v is False:
+                switch = "no-" + switch
+            params.append("--" + switch)
+            if v is not True and v is not False:
+                params.append(str(v))
+        params.extend(args)
+
+        if slots.output is None:
+            result = self.process.run_blocking(
+                self.executable, params, wait_for_s=wait_for_s
+            )
+            return result
+        else:
+            if slots.started:
+                self.process.started.connect(slots.started)
+            self.process.output.connect(slots.output)
+            if slots.finished:
+                self.process.finished.connect(slots.finished)
+            self.process.run(self.executable, params)
+
+    def install(self, packages, slots=Process.Slots(), **kwargs):
+        """Use pip to install a package or packages
+
+        If the first parameter is a string one package is installed; otherwise
+        it is assumed to be an iterable of package names.
+
+        Any kwargs are passed as command-line switches. A value of None
+        indicates a switch without a value (eg --upgrade)
+        """
+        if isinstance(packages, str):
+            return self.run(
+                "install", packages, wait_for_s=180.0, slots=slots, **kwargs
+            )
+        else:
+            return self.run(
+                "install", *packages, wait_for_s=180.0, slots=slots, **kwargs
+            )
+
+    def uninstall(self, packages, slots=Process.Slots(), **kwargs):

```

Comment:
if you make it `def uninstall(self, *packages, ...)` wouldn't it be possible to avoid the if-statement? (Same for `install` above)